src/esphome/sensor/mhz19_component.cpp

```diff
@@ -26,18 +27,59 @@ uint8_t mhz19_checksum(const uint8_t *command) {
 
 void MHZ19Component::update() {
   uint8_t response[MHZ19_RESPONSE_LENGTH];
+repeat:
   if (!this->mhz19_write_command_(MHZ19_COMMAND_GET_PPM, response)) {
     ESP_LOGW(TAG, "Reading data from MHZ19 failed!");
     this->status_set_warning();
     return;
   }
 
   if (response[0] != 0xFF || response[1] != 0x86) {
-    ESP_LOGW(TAG, "Invalid preamble from MHZ19!");
+    ESP_LOGW(TAG, "Invalid preamble from MHZ19! [%0x %0x %0x %0x %0x %0x %0x %0x %0x]",
+      response[0], response[1], response[2], response[3], response[4],
+      response[5], response[6], response[7], response[8]);
     this->status_set_warning();
+    if (++this->cleanup_rx_buffer > 2) {
+        /* try to recover and give up on it if several attemprs were no successful until the next update cycle */
+        this->cleanup_rx_buffer = 0;
+        return;
+    }
+    goto repeat;
```

Comment:
`goto` is a big anti-pattern - could this be changed into a simple for loop?

Reply:
sure  it should be possible